### PR TITLE
Support duplicating pages, connection, components

### PR DESCRIPTION
**Pull request by Janpot**
* Change delete button to menu with delete and duplicate items
* Hide menu button until hover
* Add duplicate function for pages, connections, components
* Had to rewrite the duplication function once again to allow for navigating to the newly created node afterwards.
  * This introduces the concept of DOM fragments, which can be thought of as a subtree of the dom (e.g. a page and all of its children) It is implemented using the exact same datastructure as the main `AppDom`. 
  * Fragments can be cloned and inserted in application dom. Which I implemented in this PR
  * This forms the basis for cut/copy/paste which will probably rely heavily on extracting and inserting dom fragments
* This constraint is also present on the query duplication, we will have to open the query editor after duplication. (right?)

https://user-images.githubusercontent.com/2109932/197823814-bb711a6b-b89f-4fe3-80d3-0d81a6429a8f.mov

Started on https://github.com/mui/mui-toolpad/issues/1172 (I was already halfway to adding duplication to the explorer when I noticed the ticket was about queries. But the bulk of the work is a refactoring that will be necessary for duplication of queries as well)

Fixes https://github.com/mui/mui-toolpad/issues/1173

### CODE REVIEW

**Comment by Janpot** [comment]
> Could you just clarify a bit more why fragments are needed?

To be able to navigate to the page after duplicating it we need access to the id of duplicated node. To get access to this id I split the duplication process into two: `cloneFragment`, which gives you a cloned dom with the root element being the id that we want. and then `addFragment` which inserts the cloned nodes. For when we add cut/paste we can add a `getFragment` to get a segment from the dom, followed by a `removeNode` and a `addFragment`. For this we will also need to be able to specify parent index, and not just paste at the end.

File: test/integration/duplication/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  {
+    await editorModel.openHierarchyMenu('connection');
+    const duplicateMenuItem = page.getByRole('menuitem', { name: 'Duplicate connection' });
+    await Promise.all([duplicateMenuItem.click(), page.waitForNavigation()]);
+
+    const input = page.getByLabel('base url');
+    await expect(input).toHaveValue('https://example.com/');
+
+    await editorModel.openHierarchyMenu('connection1');
```
Comment (apedroferreira) [inline test/integration/duplication/index.spec.ts:25]: Would it be worth testing that the new menu item is added after duplicating, and removed after deleting?

**Review by apedroferreira — CHANGES_REQUESTED** [review]
Looks good, I think fragments will be useful!
I found a possible issue, let me know if it makes sense.

File: packages/toolpad-app/src/toolpad/AppEditor/HierarchyExplorer/index.tsx [hunk 2]
```diff
@@ -222,8 +304,28 @@ export default function HierarchyExplorer({ appId, className }: HierarchyExplore
...
+  const handleDuplicateNode = React.useCallback(
+    (nodeId: NodeId) => () => {
+      const node = appDom.getNode(dom, nodeId);
+
+      invariant(
+        node.parentId && node.parentProp,
+        'Duplication should never be called on nodes that are not placed in the dom',
+      );
+
+      const fragment = appDom.cloneFragment(dom, node.id);
```
[the remark this answers is not in the document]
Comment (Janpot) [reply packages/toolpad-app/src/toolpad/AppEditor/HierarchyExplorer/index.tsx:316]: https://github.com/mui/mui-toolpad/pull/1210#issuecomment-1292036450

**Comment by Janpot** [comment]
> I tried duplicating "page1" and it seems that the suggested name is "page11" instead of "page2"

👍 My reasoning was that we shouldn't assume the number at the end of a name was added by us. And we should just always restart numbering. But that is probably less intuitive. I changed it to always increment if the name already starts with a number.

**Review by apedroferreira — APPROVED** [review]
Looks good, thanks for the changes!

**Pull request merged**